components/script/dom/htmlinputelement.rs

```diff
@@ -292,12 +316,71 @@ impl HTMLInputElementMethods for HTMLInputElement {
+            // On getting, it must return the current value
+            // (https://html.spec.whatwg.org/multipage/forms.html#concept-fe-value)
+            // of the element.
+            ValueMode::Value => self.textinput.borrow().get_content(),
+
+            // On getting, if the element has a value attribute, it
+            // must return that attribute's value; otherwise, it must
+            // return the string "on" (for default/on mode)
+            ValueMode::DefaultOn => {
+                let content = self.textinput.borrow().get_content();
+                if content.is_empty() {
+                    DOMString::from("on")
+                } else {
+                    content
+                }
+            },
+
+            // On getting, if the element has a value attribute, it
+            // must return that attribute's value; otherwise, it must
+            // return the empty string (for default mode) or the
+            // string "C:\fakepath\" followed by the name of the first
+            // file in the list of selected files, if any, or the
+            // empty string if the list is empty (for filename mode)
+            //
+            // TODO: append first selected filename if present
+            ValueMode::Default | ValueMode::Filename =>
+                self.textinput.borrow().get_content(),
+        }
     }
 
     // https://html.spec.whatwg.org/multipage/#dom-input-value
     fn SetValue(&self, value: DOMString) {
-        self.textinput.borrow_mut().set_content(value);
+
+        match self.get_value_mode() {
+            // On setting, it must set the element's value to the new
+            // value, set the element's dirty value flag to true,
+            // invoke the value sanitization algorithm, if the
+            // element's type attribute's current state defines one,
+            // and then, if the element has a text entry cursor
+            // position, should move the text entry cursor position to
+            // the end of the text field, unselecting any selected
+            // text and resetting the selection direction to none.
+            ValueMode::Value => self.textinput.borrow_mut().set_content(value),
+
+            // On setting, it must set the element's value attribute
+            // to the new value.
+            ValueMode::Default | ValueMode::DefaultOn => {
+                self.textinput.borrow_mut().set_content(value.clone());
+
+                // removing this breaks:
+                // Check the RadioNodeList.value on setting to 'on'
+                // The value attribute should be the first radio input element's value. expected "1" but got "2"
+                self.upcast::<Element>().set_string_attribute(&atom!("value"), value);
+            },
+
+            // On setting, if the new value is the empty string, it
+            // must empty the list of selected files; otherwise, it
+            // must throw an InvalidStateError exception.
```

Comment:
I might be misreading the spec or looking in the wrong place, but I don't see any step numbers on: https://html.spec.whatwg.org/multipage/forms.html#dom-input-value